Initial support for StsWebIdentityTokenFileCredentialsProvider

eduanb opened the pull request:
## :loudspeaker: Type of change
- [ ] Bugfix
- [ x] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
Adds support for StsWebIdentityTokenFileCredentialsProvider, a common use case in EKS.
Note: This is still WIP. I would first like feedback before adding tests, docs etc.

## :bulb: Motivation and Context
The recommended way of authenticating to AWS in EKS is using Service Accounts. See here https://docs.aws.amazon.com/eks/latest/userguide/pod-configuration.html

This PR adds native support for it. Users of spring-cloud-aws now only need to add the STS dependency and set up the required service account.  spring-cloud-aws will then automatically create a StsWebIdentityTokenFileCredentialsProvider. 

The inspiration behind this was https://github.com/awspring/spring-cloud-aws/issues/666


## :green_heart: How did you test it?
This still needs to be tested. I first want to hear initial feedback from the team if this will be a good feature to add. 

## :pencil: Checklist
- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [ ] All tests passing
- [ ] No breaking changes


## :crystal_ball: Next steps

maciejwalkowiak commented:
Also fixes #691

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -101,4 +112,18 @@ private static ProfileCredentialsProvider createProfileCredentialProvider(Profil
 
+	private static boolean shouldCreateStsIdentityTokenCredentialsProvider() {
+		return ClassUtils.isPresent(STS_WEB_IDENTITY_TOKEN_FILE_CREDENTIALS_PROVIDER, null)
+			&& StringUtils.hasText(System.getenv(AWS_ROLE_ARN_ENV_VARIABLE))
```
Do these properties make sense only as env variables or we should also include them in `CredentialsProperties`?

maciejwalkowiak reviewed (changes requested):
Thanks @eduanb for a PR! Could you please also update reference documentation?

MatejNedic commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -45,6 +49,9 @@
 public class CredentialsProviderAutoConfiguration {
 
+	public static final String STS_WEB_IDENTITY_TOKEN_FILE_CREDENTIALS_PROVIDER = "software.amazon.awssdk.services.sts.auth.StsWebIdentityTokenFileCredentialsProvider";
+	public static final String AWS_ROLE_ARN_ENV_VARIABLE = "AWS_ROLE_ARN";
```
We should modify `CredentialsProperties` with nested `StsProperties` and move all of this there.

MatejNedic commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -101,4 +112,18 @@ private static ProfileCredentialsProvider createProfileCredentialProvider(Profil
+			.stsClient(StsClient.builder().build())
+			.roleArn(System.getenv(AWS_ROLE_ARN_ENV_VARIABLE))
+			.webIdentityTokenFile(Path.of(System.getenv(AWS_WEB_IDENTITY_TOKEN_FILE_ENV_VARIABLE)))
+			.build();
```
We should add:
`.asyncCredentialUpdateEnabled(stsProperties.isAsyncCredentialsUpdate())`

So users have options to run sync or async calls.

MatejNedic commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -101,4 +112,18 @@ private static ProfileCredentialsProvider createProfileCredentialProvider(Profil
+
+	private static StsWebIdentityTokenFileCredentialsProvider createStsCredentialsProvider() {
+		return StsWebIdentityTokenFileCredentialsProvider.builder()
+			.stsClient(StsClient.builder().build())
```
Lets include `StsWebIdentityTokenFileCredentialsProvider.builder().roleSessionName()`

MatejNedic reviewed (changes requested):
Hey, @eduanb thanks for the PR!
Looks good so far. Could you apply the requested changes and updated docs, please?

eduanb commented:
Thank you for the feedback. I'm now confident that this will be a valid feature to add. 
I will work on the feedback, tests and documentation soon.

eduanb commented:
Hi @marcingrzejszczak and @MatejNedic, I've completely redone almost everything in this PR. I've added a test and documentation. 

I decided the `stsProperties` are big enough to be its own class. I've also made it use the standard `AWS_ROLE_ARN` so it could be configured with env variables or Spring Boot properties. 

I've added many debug logs and documented that the user can use them. Let me know if this is a good idea or should rather be removed.

MatejNedic commented:
Hey @eduanb , looks good, thank you so much for PR!
I did some small polish @maciejwalkowiak could you review it, please?

MatejNedic commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/StsProperties.java:
```diff
@@ -0,0 +1,105 @@
+	 * Activate {@link software.amazon.awssdk.services.sts.auth.StsWebIdentityTokenFileCredentialsProvider}
+	 * AutoConfiguration for {@link software.amazon.awssdk.crt.auth.credentials.CredentialsProvider}.
+	 */
+	private boolean enabled = false;
```
I would like to keep this since I am not a fan of autoconfiguring STS based only on if the class is on the classpath. There are multiple ways you can do STS and we should not force it.
This also helps if you want locally not to use STS, but once deployed to use it.

eduanb replied:
I agree this should be configurable. I wonder if we should make the default `true` because you already need to add the dependency and it is common for Spring Boot to auto-configure things based on dependencies. Additionally, you need to set the role and have a credentials file so I don't think it will happen accidentally locally. I also like the idea of zero-config authentication in EKS. 

That said, a false default is also fine by me.

MatejNedic replied:
I agree with all of this, but imagine someone using STS for one client and (accessKey, secretKey) for all other clients. This way he can just disable STS autoconfiguration.
Was thinking about this as well, I know it is a weird use case but I can see it happening.

maciejwalkowiak commented on docs/src/main/asciidoc/core.adoc:
```diff
@@ -61,9 +76,45 @@ If it does not serve your project needs, this behavior can be changed by setting
+----
+
+===== Configuring
+In EKS no additional configuration is required as the service account already configures the correct environment variables; however, they can be overridden.
```
We should have here a table with available properties and i don't think we need to document system properties and env variables used by AWS SDK. Maybe we can link to their official docs.

MatejNedic replied:
Agree

MatejNedic replied:
I think saying `(falls back to SDK default)`. What do you think @maciejwalkowiak

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfigurationTests.java:
```diff
@@ -89,6 +96,42 @@ void credentialsProvider_profileNameAndPathConfigured_configuresProfileCredentia
+	}
+
+	@Test
+	void credentialsProvider_stsSystemPropertiesDefault_configuresStsWebIdentityTokenFileCredentialsProvider()
```
This test is failing, for a good reason - `spring.cloud.aws.credentials.sts.enabled` has been removed.

Look another comment

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -73,6 +83,11 @@ public static AwsCredentialsProvider createCredentialsProvider(CredentialsProper
 		}
 
+		StsProperties sts = properties.getSts();
+		if (ClassUtils.isPresent(STS_WEB_IDENTITY_TOKEN_FILE_CREDENTIALS_PROVIDER, null) && sts != null) {
```
Perhaps we should not check for `sts != null` and return `StsWebIdentityTokenFileCredentialsProvider` if sts is on the classpath.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/core/CredentialsProviderAutoConfiguration.java:
```diff
@@ -101,4 +116,20 @@ private static ProfileCredentialsProvider createProfileCredentialProvider(Profil
+	 * Wrapper class to avoid {@link NoClassDefFoundError}.
+	 */
+	private static class StsCredentialsProviderFactory {
+		private static AwsCredentialsProvider create(StsProperties stsProperties, AwsRegionProvider regionProvider) {
```
If the comment above makes sense, `stsProperties` here would be `@Nullable`

maciejwalkowiak commented:
@MatejNedic please take a look at comments above. If this make sense, i have a commit ready locally

MatejNedic commented:
@maciejwalkowiak I agree with all the above.

maciejwalkowiak commented:
@MatejNedic please take another look at the recent changes. If it matches what you had in mind we can merge

MatejNedic commented:
@maciejwalkowiak looks good to me
